src/components/Table/utils/utils.tsx

```diff
@@ -0,0 +1,240 @@
+    // When using client-side sorting, we call `applyRowFn` not only during rendering, but
+    // up-front against all rows (for the currently sorted column) to determine their
+    // sort values.
+    //
+    // Pedantically this means that any table using client-side sorting should not
+    // build JSX directly in its GridColumn functions, but this overhead is especially
+    // noticeable for large/virtualized tables, so we only enforce using functions
+    // for those tables.
+    throw new Error(
+      "GridTables with as=virtual & sortable columns should use functions that return JSX, instead of JSX",
+    );
+  }
+
+  content =
+    isGridCellContent(maybeContent) && !!maybeContent.onClick
+      ? getButtonOrLink(content, maybeContent.onClick, {
+          css: Css.maxw100.lightBlue700.ta("inherit").if(style?.presentationSettings?.wrap === false).truncate.$,
+        })
+      : content;
+
+  if (content && typeof content === "string" && isHeader && canSortColumn) {
+    return <SortHeader content={content} iconOnLeft={alignment === "right"} />;
+  } else if (content && style?.presentationSettings?.wrap === false && typeof content === "string") {
+    // In order to truncate the text properly, then we need to wrap it in another element
+    // as our cell element is a flex container, which don't allow for applying truncation styles directly on it.
+    return (
+      <span css={Css.truncate.mw0.$} title={content}>
+        {content}
+      </span>
+    );
+  } else if (style.emptyCell && isContentEmpty(content)) {
+    // If the content is empty and the user specified an `emptyCell` node, return that.
+    return style.emptyCell;
+  }
+  return content;
+}
+
+export function isGridCellContent(content: ReactNode | GridCellContent): content is GridCellContent {
+  return typeof content === "object" && !!content && "content" in content;
+}
+
+const emptyValues = ["", null, undefined] as any[];
+function isContentEmpty(content: ReactNode): boolean {
+  return emptyValues.includes(content);
+}
+
+/** Return the content for a given column def applied to a given row. */
+export function applyRowFn<R extends Kinded>(
+  column: GridColumn<R>,
+  row: GridDataRow<R>,
+  api: GridTableApi<R>,
+  level: number,
+): ReactNode | GridCellContent {
+  // Usually this is a function to apply against the row, but sometimes it's a hard-coded value, i.e. for headers
+  const maybeContent = column[row.kind];
+  if (typeof maybeContent === "function") {
+    // Auto-destructure data
+    return (maybeContent as Function)((row as any)["data"], { row: row as any, api, level });
+  } else {
+    return maybeContent;
+  }
+}
+
+/**
+ * Filters rows given a client-side text `filter.
+ *
+ * Ensures parent rows remain in the list if any children match the filter.
+ *
+ * We return a copy of `[Parent, [Child]]` tuples so that we don't modify the `GridDataRow.children`.
+ */
+export function filterRows<R extends Kinded>(
```

Comment:
Question: Is this really called by multiple places now? Otherwise would wonder if it just should live in whatever file is calling it, same question for the others, my initial guess is that most? of them are only used by one file so imo would be cleaner to just keep them next to their main/only caller.

Reply:
Nope, just in one. Moved it.